pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
 		return reconcile.Result{}, err
 	}
 
-	if !isHealthy(node) {
-		if unhealthyForTooLong(node) {
-			// delete
+	cmUnhealthConditions, err := getUnhealthConditionsConfigMap(r)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	nodeUnhealthConditions, err := conditions.GetNodeUnhealthyConditions(node, cmUnhealthConditions)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var result *reconcile.Result
+	for _, c := range nodeUnhealthConditions {
+		nodeCondition := conditions.GetNodeCondition(node, c.Name)
+		// skip when current node condition is different from the one reported in the config map
+		if nodeCondition == nil || !isConditionsStatusesEqual(nodeCondition, &c) {
+			continue
+		}
+
+		conditionTimeout, err := time.ParseDuration(c.Timeout)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// apply remediation logic, if at least one condition last more than specified timeout
+		if unhealthyForTooLong(nodeCondition, conditionTimeout) {
 			glog.Infof("machine %s has been unhealthy for too long, deleting", machine.Name)
 			if err := r.client.Delete(context.TODO(), machine); err != nil {
 				glog.Errorf("failed to delete machine %s, requeuing referenced node", machine.Name)
 				return reconcile.Result{}, err
 			}
 			return reconcile.Result{}, nil
-		} else {
-			currentTime := time.Now()
-			lastTimeHealthy := lastTransitionTime(node, healthCriteria)
-			durationUnhealthy := currentTime.Sub(lastTimeHealthy.Time)
-			glog.Warningf("machine %s have had node %s unhealthy for %s. Requeuing...", machine.Name, node.Name, durationUnhealthy.String())
-			return reconcile.Result{Requeue: true, RequeueAfter: remediationWaitTime}, nil
 		}
+
+		now := time.Now()
+		durationUnhealthy := now.Sub(nodeCondition.LastTransitionTime.Time)
+		glog.Warningf(
+			"machine %s have unhealthy node %s with condition %s for %s. Requeuing...",
+			machine.Name,
+			node.Name,
+			nodeCondition.Type,
+			durationUnhealthy.String(),
+		)
+		result = &reconcile.Result{Requeue: true, RequeueAfter: conditionTimeout}
 	}
+
+	// requeue
+	if result != nil {
+		return *result, nil
+	}
+
 	glog.Infof("No remediaton action was taken. Machine %s with node %v is healthy", machine.Name, node.Name)
 	return reconcile.Result{}, nil
 }
 
+func getUnhealthConditionsConfigMap(r *ReconcileMachineHealthCheck) (*corev1.ConfigMap, error) {
+	cmUnhealthConditions := &corev1.ConfigMap{}
+	cmKey := types.NamespacedName{
+		Name:      healthcheckingv1alpha1.ConfigMapNodeUnhealthConditions,
+		Namespace: r.namespace,
+	}
+	err := r.client.Get(context.TODO(), cmKey, cmUnhealthConditions)
+	if err != nil {
+		// Error reading the object - requeue the request
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+
+		// creates dummy config map with default values if it does not exist
+		cmUnhealthConditions, err = conditions.CreateDummyUnhealthyConditionsConfigMap()
```

Comment:
@enxebre this needs to be documented though so a user is not surprised in case he/she deletes the configmap (or all producers of it)

Reply:
@ingvagabund it is another question, do I need to add documentation somewhere?

Reply:
@cynepco3hahue can you include it here https://github.com/openshift/machine-api-operator#machine-healthcheck-controller